src/trait_/common_ord.rs

```diff
@@ -106,98 +113,77 @@ pub fn build_ord_signature(item: &Item, trait_: &DeriveTrait, body: &TokenStream
+								)
+							},
+						};
+
+						quote! {
+							#incomparable
+
+							let __self_disc = ::core::mem::discriminant(self);
+							let __other_disc = ::core::mem::discriminant(__other);
+
+							if __self_disc == __other_disc {
+								#body_equal
+							} else {
+								#body_else
+							}
 						}
 					}
 				}
 				// Safe implementation when not on nightly.
 				#[cfg(all(not(feature = "nightly"), feature = "safe"))]
 				{
-					use syn::PatOr;
-
-					let mut less = quote! { ::core::cmp::Ordering::Less };
-					let mut greater = quote! { ::core::cmp::Ordering::Greater };
-
-					// Add `Option` to `Ordering` if we are implementing `PartialOrd`.
-					if let PartialOrd = trait_ {
-						less = quote! { ::core::option::Option::Some(#less) };
-						greater = quote! { ::core::option::Option::Some(#greater) };
-					}
-
-					let mut different = Vec::with_capacity(variants.len());
-					let variants: Vec<_> =
-						variants.iter().filter(|v| !v.is_incomparable()).collect();
-
-					// Build separate `match` arms to compare different variants to each
-					// other. The index for these variants is used to determine which
-					// `Ordering` to return.
-					for (index, variant) in variants.iter().enumerate() {
-						let pattern = &variant.self_pattern();
-
-						// The first variant is always `Less` then everything.
-						if index == 0 {
-							different.push(quote! {
-								#pattern => #less,
-							})
-						}
-						// The last variant is always `Greater` then everything.
-						else if index == variants.len() - 1 {
-							different.push(quote! {
-								#pattern => #greater,
-							})
+					let body_else = match discriminant {
+						Discriminant::Single => {
+							unreachable!("we should only generate this code with multiple variants")
 						}
-						// Any variant between the first and last.
-						else {
-							// Collect all variants that are `Less`.
-							let cases = variants
-								.iter()
-								.enumerate()
-								.filter(|(index_other, _)| *index_other < index)
-								.map(|(_, variant_other)| {
-									variant_other.other_pattern_skip().clone()
-								})
-								.collect();
-
-							// Build one match arm pattern with all variants that are `Greater`.
-							let pattern_less = PatOr {
-								attrs: Vec::new(),
-								leading_vert: None,
-								cases,
-							};
-
-							// All other variants are `Less`.
-							different.push(quote! {
-								#pattern => match __other {
-									#pattern_less => #greater,
-									_ => #less,
-								},
-							});
+						Discriminant::UnitDefault => quote! {
+							#path::#method(
+								self as isize,
+								__other as isize,
+							)
+						},
+						Discriminant::UnitRepr(repr) => quote! {
+							#path::#method(
+								self as #repr,
+								__other as #repr,
+							)
+						},
+						Discriminant::Default | Discriminant::Repr(_) => {
+							build_recursive_order(trait_, variants, &incomparable)
 						}
```

Comment:
the only difference between safe and unsafe seems to be the handling of `Repr`, maybe it would make more sense to merge the two and only differentiate in that branch

Reply:
Done.